add new `trim_array_spaces` fixer

jaredh159 opened the pull request:
Single line arrays should be formatted like function/method argments without leading or trailing spaces.

After switching to PSR-2 I noticed that my old coding style for single-line arrays didn't seem to match the overall aesthetic of PSR-2.  I created this fixer to convert arrays like:

`$foo = array( 'foo', 'bar' );`

to this style:

`$foo = array('foo', 'bar');`

GrahamCampbell commented:
What if an array contains an anonymous function, or a method call? We need tests to show the brackets are not moved. We also need tests for short array syntax.

GrahamCampbell commented:
We also need tests to deal with comments correctly.

GrahamCampbell commented:
Should this be in the symfony level @keradus?

keradus commented:
I agree

keradus commented:
We also need tests for array inside array when array inside should be fixed but not array outside, opposite case, both to be fixed and both to not be fixed.

jaredh159 commented:
I added a bunch of requested tests.  I wasn't sure what kind of tests you were looking for regarding comments, @GrahamCampbell, can you give me a few pointers there?  Also, per the discussion regarding symfony level, do I need to move the fixer out of Contrib eventually?

GrahamCampbell commented:
> Also, per the discussion regarding symfony level, do I need to move the fixer out of Contrib eventually?

I think so yeh. :)

GrahamCampbell commented:
> can you give me a few pointers there?

Cases like:

```php

$foo = array( 'foo' =>  [    ] );

bar([ /* empty array */        ]);
```

GrahamCampbell commented:
Also, cases like:

```php
$foo = array($bar[  4 ] );
```

We'd need to prove that the braces not in the context for the array are unmodifed.

stof commented:
@GrahamCampbell In Symfony, we indeed don't use spaces in these places, so this fixer is legitimate at the symfony level

jaredh159 commented:
added requested tests (I think!) and moved fixer into Symfony level. let me know if there's anything else I need to refine or modify.

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/TrimArraySpacesFixer.php:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function getDescription()
```
Please put the fix method first to match all the other fixers.

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/TrimArraySpacesFixer.php:
```diff
@@ -0,0 +1,75 @@
+     * @param Tokens $tokens
+     * @param int    $index
+     */
+    private function fixArray(Tokens $tokens, $index)
```
Should this be static?

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/TrimArraySpacesFixer.php:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        if ($tokens[$startIndex + 1]->isWhitespace()) {
+            $tokens[$startIndex + 1]->clear();
```
Don't call this twice. Cache the token by setting it to a local variable first.

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/TrimArraySpacesFixer.php:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        if ($tokens[$endIndex - 1]->isWhitespace()) {
+            $tokens[$endIndex - 1]->clear();
```
Same here.

GrahamCampbell commented on Symfony/CS/Tests/Fixer/Symfony/TrimArraySpacesFixerTest.php:
```diff
@@ -0,0 +1,212 @@
+
+            array(
+                "<?php \$foo = [/* empty array */];",
+                "<?php \$foo = [ /* empty array */ ];",
```
We need a more complex mix of a load of different stuff. Nested arrays, with different syntax, and as much crazy stuff as possible. This very separate test cases don't quite cut it. We need some more "real life" cases.

jaredh159 commented:
Tried to address the concerns you raised @GrahamCampbell.

GrahamCampbell commented on Symfony/CS/Tests/Fixer/Symfony/TrimArraySpacesFixerTest.php:
```diff
@@ -0,0 +1,295 @@
+
+            array(
+                "<?php \$foo = [function() {return 'foo';}];",
+                "<?php \$foo = [ function() {return 'foo';} ];",
```
can we include some more bad spacing to check only the correct stuff is changed? Same for the other test cases.

jaredh159 replied:
I'm afraid I don't follow what you mean. Can you give me an example or two?

GrahamCampbell replied:
```php
"<?php \$foo = [   function   (    ) {        return 'foo ';  }   ];",
```

GrahamCampbell replied:
We should make sure only the spaces in the correct content are removed.

GrahamCampbell replied:
Also, maybe some tests with tabs as well as spaces would be a good idea. :)

jaredh159 commented:
Thanks @GrahamCampbell that helped.  Added some tests per your recommendation.

GrahamCampbell commented:
Alothug we have tests to check tabs in the incorrect context aren't changed, we don't have any tests that look like this; `"array(\t'test')"`, where the tab should be removed by the fixer.

GrahamCampbell commented:
Otherwise, this is looking pretty good now.

jaredh159 commented:
@GrahamCampbell yup, thanks, hadn't thought of that. Tests added.

stof commented on Symfony/CS/Tests/Fixer/Symfony/TrimArraySpacesFixerTest.php:
```diff
@@ -0,0 +1,355 @@
+
+            array(
+                "<?php \$foo = array();",
+                "<?php \$foo = array( );",
```
Use single quoted strings, which will avoid the need to escape ``$`` here

stof commented on Symfony/CS/Tests/Fixer/Symfony/TrimArraySpacesFixerTest.php:
```diff
@@ -0,0 +1,355 @@
+
+            array(
+                "<?php \$foo = ['foo'];",
+                "<?php \$foo = [ 'foo' ];",
```
any testcase using thr short array syntax should not run on PHP 5.3, but only on PHP 5.4+ (similar to the way you run some tests only on 5.5+)

jaredh159 replied:
@stof I'm not sure I follow why. This Fixer doesn't change short to long or long to short.  It handles both.  It can run on 5.3+.  I looked at some of the other test cases that involve both types of arrays and they don't split them out into different test methods requiring different versions. Am I missing something? I'm happy to do it if necessary, I just am not understanding why.

stof replied:
the test using the short array syntax will not work on PHP 5.3 because it will throw a parse error when tokenizing the code. this is why it should be skipped

stof replied:
hmm, actually, it is really weird. The test seems to be passing.

jaredh159 replied:
yeah, because it's not executing the code, it's just fixing it.  So, PHP 5.3 can still fix PHP 5.4 code. At least that's the way I think of it.

keradus commented:
Please add a test with a comma after last element

keradus commented on Symfony/CS/Fixer/Symfony/TrimArraySpacesFixer.php:
```diff
@@ -0,0 +1,79 @@
+        $startIndex = $index;
+
+        if ($tokens[$startIndex]->isGivenKind(T_ARRAY)) {
+            $startIndex = $tokens->getNextTokenOfKind($startIndex, array('('));
```
Use `getNextMeaningfulToken`, then you won't need to pass 2nd argument

keradus commented:
Please, could you:
- rebase, to get rid of that ugly merge commit
- squash all commit of yours ?

keradus commented:
Please see https://github.com/jaredh159/PHP-CS-Fixer/pull/2

jaredh159 commented:
@keradus I rebased and squashed as requested. However, there was a failure in Travis.  It doesn't look like it's from the code added in this branch. That said, this is my first time ever rebasing or squashing, so there's a very real possibility that I messed something up.  Sorry if I did.

GrahamCampbell commented:
You need to retain @keradus's authorship on the commit.

GrahamCampbell commented:
Also, yeh, you did mess it up, lol. You rebased against the master instead of the 1.6 branch.

jaredh159 commented:
@GrahamCampbell argh, sorry. I'm such a noob. I'll try to sort it out and do it again right.  Thanks for bearing with me. I'll squash my commits and leave keradus' as well.

keradus commented:
try this:
```bash
# remove old branch:
git branch -D 1.6-trim-single-line-array-whitespace

# recreate branch as copy of 1.6:
git fetch upstream
git checkout upstream/1.6
git checkout -b 1.6-trim-single-line-array-whitespace

# apply missing commits:
git cherry-pick 33d99d4b0d931b84609cb345fed372caaed1e039 --no-commit
git commit -m "Added TrimArraySpacesFixer"
git cherry-pick 05f325552ae1b62285ed5e25a2b17d337d7f86dd

# update the origin
git push origin 1.6-trim-single-line-array-whitespace -f
```

keradus commented:
To help things out: recreated into #1090 .
